### PR TITLE
feat: add MediaSession API

### LINKED ISSUE #32177: [Feature Request]: MediaSession API bindings
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

The [MediaSession API](https://developer.mozilla.org/en-US/docs/Web/API/Media_Session_API) provides a way for web applications to communicate to the browser client which media-related actions it supports. In response, a browser client can display relevant controls pertaining to an active media session. This is used for media playback or video conferencing applications. More examples and info on the API can be found at https://web.dev/media-session/

While web applications are able to register MediaSession action handlers, it's not possible to know which are registered from the browser/main process. Applications could use this information to show controls integrated more closely with the OS.

### Proposed Solution

The [MediaSession interface](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/media_session.h) exists in Chromium for each WebContents. It can be used to control an active media session and receive updates on when actions are registered. Providing V8 bindings for this interface would allow Electron applications to be aware of such media actions and to build their own interfaces on top of them.

```ts
interface WebContents {
  /** Lazy property getter for the corresponding MediaSession. */
  get mediaSession(): MediaSession;
}

/**
 * Class for binding content::MediaSession.
 * 
 * https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/media_session.h
 * https://source.chromium.org/chromium/chromium/src/+/main:services/media_session/public/mojom/media_session.mojom
 */
interface MediaSession {
  // MediaSession mojo interface bindings
  suspend(): void;
  resume(): void;
  startDucking(): void;
  stopDucking(): void;
  getMediaSessionInfo(): Promise<MediaSessionInfo>;
  previousTrack(): void;
  nextTrack(): void;
  skipAd(): void;
  seek(seekTime: number): void;
  stop(): void;
  getMediaImageBitmap(): Promise<NativeImage>;
  seekTo(seekTime: number): void;
  scrubTo(seekTime: number): void;
  toggleMicrophone(): void;
  toggleCamera(): void;
  hangUp(): void;
  raise(): void;
  setMute(muted: boolean): void;

  // MediaSessionObserver bindings
  on('info-changed', listener: (info: MediaSessionInfo) => void): void;
  on('metadata-changed', listener: (metadata: MediaMetadata) => void): void;
  on('actions-changed', listener: (actions: MediaSessionAction[]) => void): void;
  on('images-changed', listener: (images: Map<MediaSessionImageType, NativeImage>) => void): void;
  on('position-changed', listener: (position: MediaPosition) => void): void;
}

```

### Alternatives Considered

None

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by samuelmaddock**
#### Description of Change

resolves #32177 

This is an initial pass at introducing a `MediaSession` API to the main process to interact with its renderer counterpart. The latter is the [W3C Media Session API](https://www.w3.org/TR/mediasession/).

The goal is to further allow Electron applications to customize media playback controls, beyond what's already provided by Chromium (see [web.dev article for examples](https://web.dev/media-session/)).

Please see #32177 for more details about motivation and use cases.

I've left TODO comments in the code to point attention to remaining work. At a high-level, here's a known list of what's left:
- Add bindings for invoking more action handlers (seek, skip, toggle camera/mic)
- Implement remaining media events (metadata, playback position)
- Design and serialize event details
  - Might be worth using lazy property getters for media thumbnail metadata updates
- Fix DCHECK crash when calling action handlers with inactive audio session

I'd like to get feedback on the API's current state and whether there are any major design changes which should be made prior to completing the above list of tasks.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added MediaSession API to the main process for accessing media metadata and action handlers.

### CODE REVIEW

File: shell/browser/api/electron_api_media_session.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      v8::Isolate* isolate) override;
+  const char* GetTypeName() override;
+
+  // disable copy
+  MediaSession(const MediaSession&) = delete;
+  MediaSession& operator=(const MediaSession&) = delete;
+
+ protected:
+  MediaSession(v8::Isolate* isolate, content::MediaSession* media_session);
+  ~MediaSession() override;
```
Comment (RaisinTen) [inline shell/browser/api/electron_api_media_session.h:46]: Why not make these private? Are we planning on creating subclasses of this later on?

File: shell/browser/api/electron_api_media_session.cc [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    // media_session::mojom::MediaSessionInfo::SessionState::kDucking);
+
+    std::string playback_state;
+    switch (val->playback_state) {
+      case media_session::mojom::MediaPlaybackState::kPlaying:
+        playback_state = "playing";
+        break;
+      case media_session::mojom::MediaPlaybackState::kPaused:
+        playback_state = "paused";
+        break;
```
[the remark this answers is not in the document]
Comment (samuelmaddock) [reply shell/browser/api/electron_api_media_session.cc:74]: By leaving out the `default` case, enumerating all possible values will be enforced at compile time. This is actually the behavior I wanted for `media_session::mojom::MediaSessionAction` as well so I'll plan to remove the `default` case there too.

File: shell/browser/api/electron_api_media_session.h (hunk 1 above)
Comment (samuelmaddock) [reply shell/browser/api/electron_api_media_session.h:46]: This file initially started by duplicating https://github.com/electron/electron/blob/main/shell/browser/api/electron_api_debugger.h

I'm not too sure whether it's necessary or not. I see the same thing in [gin::Wrappable](https://source.chromium.org/chromium/chromium/src/+/main:gin/wrappable.h;l=105-107;drc=68c0798d46c47105ffdbf40c58bb63153a148116).

@deepak1556 Do you happen to recall the reasoning for setting this to protected? I realize the `Debugger` API was introduced quite a while ago now. 😅

File: shell/browser/api/electron_api_media_session.cc [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    const absl::optional<media_session::MediaMetadata>& metadata_mojo) {
+  // TODO(samuelmaddock): emit event
+}
+
+void MediaSession::MediaSessionActionsChanged(
+    const std::vector<media_session::mojom::MediaSessionAction>& actions) {
+  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  gin_helper::Dictionary details = gin_helper::Dictionary::CreateEmpty(isolate);
+  // TODO(samuelmaddock): should this property be a Set<string>?
```
[the remark this answers is not in the document]
Comment (samuelmaddock) [reply shell/browser/api/electron_api_media_session.cc:118]: Discussed in Slack and settled on just keeping `Array<string>` for consistency and reducing conversion cost.

File: shell/browser/api/electron_api_media_session.cc [hunk 4]
```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2022 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/api/electron_api_media_session.h"
+
+#include <string>
+#include <vector>
+
```
[the remark this answers is not in the document]
Comment (samuelmaddock) [reply shell/browser/api/electron_api_media_session.cc:9]: https://github.com/electron/electron/pull/32848#discussion_r805065473

File: shell/browser/api/electron_api_media_session.h (hunk 1 above)
Comment (RaisinTen) [reply shell/browser/api/electron_api_media_session.h:46]: Ah, I believe the Debugger class ctor can be turned private too. However, Wrappable is subclassed at multiple places, like https://github.com/electron/electron/blob/84cf685e761b2dd22f064e0dbf33b290bbec5fc0/shell/browser/api/electron_api_auto_updater.h#L20 for example, so this wouldn't be applicable for the Wrappable class.

File: docs/api/media-session.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  }
+)
+
+let playbackState
+
+win.webContents.mediaSession.on(
+  'info-changed',
+  (event, details) => {
+    // Save last known playback state.
+    playbackState = details.playbackState
```
Comment (nornagon) [inline docs/api/media-session.md:41]: `const`?

**Comment by nornagon** [comment]
API LGTM

**Comment by geroale** [comment]
Any chance to merge it? @nornagon @RaisinTen

**Comment by samuelmaddock** [comment]
> Any chance to merge it? @nornagon @RaisinTen

This PR needs to be rebased and the remaining todos in the description need to be completed. I'm not sure when I would get around to this, but I'd be open to adding you as a contributor to my fork if you're eager to take on this work!

**Pull request closed without merging**